### PR TITLE
More Ruby-eqsue interface

### PR DESCRIPTION
**Pull request by marcinwyszynski**
There are two changes in this PR.

First change implements `respond_to_missing?` which is expected by a ton of Ruby APIs, and is a [good thing](https://robots.thoughtbot.com/always-define-respond-to-missing-when-overriding) to do. The implementation shares a ton of code with `method_missing` and I briefly thought about factoring it out to a shared function but I decided against it since it would create an abstraction that is hard to follow.

Second change makes `==` a bit more lenient. The comparison will now return `false` instead of failing if a message is compared to something with a different type. This is how Ruby APIs generally work:

```
$ irb
irb(main):001:0> Time.now == nil
=> false
irb(main):002:0> 7 == nil
=> false
irb(main):003:0> 7 == "7"
=> false
irb(main):004:0> Time.now == "bacon"
=> false
```

### CODE REVIEW

**Comment by haberman** [comment]
ok to test

**Comment by haberman** [comment]
Sorry for the delay. This change looks good, but please fix this warning before I merge it:

```
../../../../ext/google/protobuf_c/message.c: In function 'Message_eq':
../../../../ext/google/protobuf_c/message.c:348:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   MessageHeader* self;
   ^
```

**Comment by marcinwyszynski** [comment]
@haberman - fair point, I missed that. Fixed now, PTAL.

**Comment by marcinwyszynski** [comment]
@haberman any chance this change can be released on rubygems some time soon? A library I want to publish depends on this behavior.
